runtime/vm/valueTypeHelpers.hpp

```diff
@@ -37,14 +43,59 @@ class ValueTypeHelpers {
 	 * Function members
 	 */
 private:
+	static bool
+	isSubstitutable(J9VMThread *currentThread, MM_ObjectAccessBarrierAPI objectAccessBarrier, j9object_t lhs, j9object_t rhs, UDATA startOffset, J9Class *clazz);
 
+	static VMINLINE bool
+	checkDoubleEquality(U_64 a, U_64 b)
+	{
+		bool result = false;
+
+		if (a == b) {
+			result = true;
+		} else if (IS_NAN_DBL(*(jdouble*)&a) && IS_NAN_DBL(*(jdouble*)&b)) {
+			result = true;
+		}
+		return result;
+	}
+
+	static VMINLINE bool
+	checkFloatEquality(U_32 a, U_32 b)
+	{
+		bool result = false;
+
+		if (a == b) {
+			result = true;
+		} else if (IS_NAN_SNGL(*(jfloat*)&a) && IS_NAN_SNGL(*(jfloat*)&b)) {
+			result = true;
+		}
+		return result;
+	}
 protected:
 
 public:
-	static bool
-	isSubstitutable(J9VMThread *currentThread, j9object_t lhs, j9object_t rhs, UDATA startOffset, J9Class *clazz)
+	static VMINLINE bool
+	acmp(J9VMThread *currentThread, MM_ObjectAccessBarrierAPI objectAccessBarrier, j9object_t lhs, j9object_t rhs)
 	{
-		return false;
+#if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
+		bool acmpResult = false;
```

Comment:
Assign inital value to (rhs == lhs) and test inside the ifdef, resulting in a single return at the end.